spynnaker/pyNN/models/neural_projections/connectors/index_based_probability_connector.py

```diff
@@ -0,0 +1,190 @@
+from spinn_utilities.overrides import overrides
+from spynnaker.pyNN.utilities import utility_calls
+from .abstract_connector import AbstractConnector
+from spinn_utilities.safe_eval import SafeEval
+import logging
+import numpy
+import math
+
+# support for arbitrary expression for the indices
+from numpy import arccos, arcsin, arctan, arctan2, ceil, cos
+from numpy import cosh, exp, fabs, floor, fmod, hypot, ldexp
+from numpy import log, log10, modf, power, sin, sinh, sqrt
+from numpy import tan, tanh, maximum, minimum, e, pi
+
+logger = logging.getLogger(__name__)
+_index_expr_context = SafeEval(math, numpy, arccos, arcsin, arctan, arctan2,
+                               ceil, cos, cosh, exp, fabs, floor, fmod, hypot,
+                               ldexp, log, log10, modf, power, sin, sinh, sqrt,
+                               tan, tanh, maximum, minimum, e=e, pi=pi)
+
+
+class IndexBasedProbabilityConnector(AbstractConnector):
+    """ Make connections using a probability distribution which varies
+        dependent upon the indices of the pre- and post-populations.
+    """
+
+    __slots = [
+        "_allow_self_connections",
+        "_index_expression",
+        "_probs"]
+
+    def __init__(
+            self, index_expression, allow_self_connections=True, rng=None,
+            safe=True, callback=None, verbose=False):
+        """
+
+        :param `string` index_expression:
+            the right-hand side of a valid python expression for
+            probability, involving the indices of the pre and post populations,
+            that can be parsed by eval(), that computes a probability dist.
+        :param `bool` allow_self_connections:
+            if the connector is used to connect a
+            Population to itself, this flag determines whether a neuron is
+            allowed to connect to itself, or only to other neurons in the
+            Population.
+        """
+        super(IndexBasedProbabilityConnector, self).__init__(safe, verbose)
+        self._index_expression = index_expression
+        self._allow_self_connections = allow_self_connections
+
+        self._probs = None
+
+    def _update_probs_from_index_expression(self, pre_vertex_slice,
+                                            post_vertex_slice):
+        # note: this only needs to be done once
+        if self._probs is None:
+            # numpy array of probabilities using the index_expression
+            self._probs = numpy.array([[_index_expr_context.eval(
+                self._index_expression, i=i, j=j)
+                for j in range(self._n_post_neurons)]
+                for i in range(self._n_pre_neurons)])
+
+    @overrides(AbstractConnector.get_delay_maximum)
+    def get_delay_maximum(self):
```

Comment:
I am guessing this should call the update function above, to ensure self._probs is filled in...

Reply:
Agreed.